### PR TITLE
Composite resource fixes

### PR DESCRIPTION
**Pull request by pkdash**

### CODE REVIEW

**Comment by pkdash** [comment]
@hyi If you can review this PR before my other PR (unit tests for view functions) that would be great. Thanks.

File: hs_core/hydroshare/utils.py [hunk 1]
```diff
@@ -800,8 +800,19 @@ def prepare_resource_default_metadata(resource, metadata, res_title):
...
+    # only add the resource creator as the creator for metadata if there is not already
+    # creator data in the metadata object
+    creator_data_exists = False
+    for element in metadata:
+        # here k is the name of the element
+        # v is a dict of all element attributes/field names and field values
+        k, v = element.items()[0]
+        if k.lower() == 'creator':
+            creator_data_exists = True
+            break
```
Comment (hyi) [inline hs_core/hydroshare/utils.py]: @pkdash Wondering whether it is possible to get all keys from metadata, then use ```'creator' in metadata_keys``` to see whether creator metadata element exists. That logic may be simpler.

File: hs_file_types/views.py [hunk 2]
```diff
@@ -481,12 +481,15 @@ def get_metadata(request, hs_file_type, file_type_id, metadata_mode):
...
+        if metadata_mode == 'view':
+            metadata = logical_file.metadata.get_html()
+        else:
+            metadata = logical_file.metadata.get_html_forms()
+        ajax_response_data = {'status': 'success', 'metadata': metadata}
+    except Exception as ex:
+        ajax_response_data = {'status': 'error', 'message': ex.message}
 
-    ajax_response_data = {'status': 'success', 'metadata': metadata}
     return JsonResponse(ajax_response_data, status=status.HTTP_200_OK)
```
Comment (hyi) [inline hs_file_types/views.py]: @pkdash It appears here that even if there is an exception that causes a status of error, the return of JsonResponse() still returns with a HTTP_200_OK status which means success. I guess it depends on how the ajax request responds to this response, but it's probably better and more reasonable locally to return a failed status in JsonResponse.

File: hs_geo_raster_resource/models.py [hunk 3]
```diff
@@ -65,6 +65,7 @@ def create(cls, **kwargs):
                 if value_item not in value_arg_dict:
                     raise ValidationError("For coverage of type 'box' values for one or more "
                                           "bounding box limits or 'units' is missing.")
+                    # TODO: Pabitra - validate also the values of the bounding box
```
Comment (hyi) [inline hs_geo_raster_resource/models.py]: @pkdash Just curious, thought the bounding box is extracted from the tif file, so what is the idea for validating the values of the bounding box in your TODO comment?

File: hs_geo_raster_resource/models.py [hunk 4]
```diff
@@ -130,25 +131,29 @@ def add_to_xml_container(self, container):
...
-    def get_html_form(self, resource):
-        """Generates html form code for this metadata element so that this element can be edited"""
+    @classmethod
+    def get_html_form(cls, resource, element=None, allow_edit=True, file_type=False):
+        """Generates html form code for an instance of this metadata element so
+        that this element can be edited"""
 
         from .forms import OriginalCoverageSpatialForm
 
         ori_coverage_data_dict = dict()
```
Comment (hyi) [inline hs_geo_raster_resource/models.py]: @pkdash better replace dict() with {} due to performance reasons. See http://stackoverflow.com/questions/664118/whats-the-difference-between-dict-and for details.

File: theme/static/js/hs_file_browser.js [hunk 5]
```diff
@@ -466,6 +466,22 @@ function showFileTypeMetadata(){
      // Wait for the asynchronous calls to finish to get new folder structure
      $.when.apply($, calls).done(function (result) {
          var json_response = JSON.parse(result);
+         if(json_response.status === 'error'){
```
Comment (hyi) [inline theme/static/js/hs_file_browser.js]: add a space before ```{```

**Comment by hyi** [comment]
@pkdash I am done with code review and have left a few comments.

File: hs_file_types/views.py (hunk 2 above)
Comment (pkdash) [inline hs_file_types/views.py]: @hyi Javscript only looks for the value of the key ```status``` to take action accordingly. I had difficulty in handling the error if status code was anything other than 200. So I have only the success handler in JS that process the response from the ajax call.

File: hs_geo_raster_resource/models.py (hunk 3 above)
Comment (pkdash) [inline hs_geo_raster_resource/models.py]: @hyi It is not needed. There was some issues with the extraction code where this element was getting created with no data. I have now fixed the extraction code. So validation here is not needed. Will remove the TODO comment.

File: hs_geo_raster_resource/models.py (hunk 4 above)
Comment (pkdash) [inline hs_geo_raster_resource/models.py]: Well PyCharm suggested me to do this way when I had what you are suggesting. I guess I can ignore PyCharm in this case.

File: theme/static/js/hs_file_browser.js (hunk 5 above)
Comment (pkdash) [inline theme/static/js/hs_file_browser.js]: Fixed.

File: hs_geo_raster_resource/models.py (hunk 4 above)
Comment (pkdash) [inline hs_geo_raster_resource/models.py]: Fixed.

File: hs_core/hydroshare/utils.py (hunk 1 above)
Comment (pkdash) [inline hs_core/hydroshare/utils.py]: Fixed.

**Comment by pkdash** [comment]
@hyi Have addressed your comments. See if you can +1.

**Comment by hyi** [comment]
+1

**Pull request merged**